src/esm_parser/provenance.py

```diff
@@ -235,12 +263,11 @@ def set_provenance(self, provenance):
             given, the ``dictionary`` takes the value of ``self``. Only for recursion
             within nested ``DictWithProvenance``, do not use it outside of this method.
         """
-
         for key, val in self.items():
             if isinstance(val, dict):
-                self[key] = DictWithProvenance(val, provenance)
-            if isinstance(val, list):
-                self[key] = ListWithProvenance(val, provenance)
+                DictWithProvenance.set_provenance(val, provenance)
+            elif isinstance(val, list):
+                ListWithProvenance.set_provenance(val, provenance)
```

Comment:
What you are doing here is using a method of a class without having an object instanciated. I don't think that's what you want to do here.

Instead I would suggest that you first instantiate the object and then you set the provenance:
```suggestion
                self[key] = DictWithProvenance(val, None)
                self[key].set_provenance(val, provenance)
            elif isinstance(val, list):
                self[key] = ListWithProvenance(val, None)
                self[key].set_provenance(val, provenance)

```

Reply:
Implemented this with small changes.